pkg/operator/awsTagController.go

```diff
@@ -0,0 +1,324 @@
+		context.Background(),
+		defaults.InfrastructureResourceName,
+		metav1.GetOptions{},
+	)
+	if err != nil {
+		klog.Errorf("failed to fetch Infrastructure resource: %v", err)
+		return err
+	}
+
+	// tags deletion is not supported. Should the user remove it from
+	// PlatformStatus will be looked up for retaining the tag
+	infraTagSet := make(map[string]string)
+	mergePlatformStatusTags(infra, infraTagSet)
+	klog.V(5).Infof("tags read from Infrastructure resource: %v", infraTagSet)
+
+	// Create a driver with the current configuration
+	ctx := context.Background()
+	driver := s3.NewDriver(ctx, cr.Spec.Storage.S3, &c.listers.StorageListers, c.featureGateAccessor)
+
+	s3TagSet, err := driver.GetStorageTags()
+	if err != nil {
+		klog.Errorf("failed to fetch storage tags: %v", err)
+		return err
+	}
+	klog.V(5).Infof("tags read from storage resource: %v", s3TagSet)
+
+	tagUpdatedCount := compareS3InfraTagSet(s3TagSet, infraTagSet)
+	if tagUpdatedCount > 0 {
+		if err := driver.PutStorageTags(s3TagSet); err != nil {
+			klog.Errorf("failed to update/delete tagset of %s s3 bucket: %v", driver.ID(), err)
+			c.event.Warningf("UpdateAWSTags",
+				"Failed to update/delete tagset of %s s3 bucket", driver.ID())
+		}
+		klog.Infof("successfully updated/deleted %d tags, tagset: %+v", tagUpdatedCount, s3TagSet)
+		c.event.Eventf("UpdateAWSTags",
+			"Successfully updated/deleted tagset of %s s3 bucket", driver.ID())
+	}
+
+	return nil
+}
+
+// mergePlatformStatusTags is for reading and merging user tags present in both
+// Platform Status of Infrastructure config.
+// There could be scenarios(upgrade, user deletes) where user tags could be missing
+// from the Platform Status, hence using Status too to avoid said scenarios.
+// If a tag exists in Status.
+func mergePlatformStatusTags(infra *configv1.Infrastructure, infraTagSet map[string]string) {
+	for _, statusTags := range infra.Status.PlatformStatus.AWS.ResourceTags {
+		if err := validateUserTag(statusTags.Key, statusTags.Value); err != nil {
+			klog.Warningf("validation failed for tag(%s:%s): %v", statusTags.Key, statusTags.Value, err)
+			continue
+		}
+		infraTagSet[statusTags.Key] = statusTags.Value
+	}
+}
+
+// compareS3InfraTagSet is for comparing the tags obtained from S3 bucket and Infrastructure CR
+// to find if any new tags have been deleted, added or existing tags modified.
+func compareS3InfraTagSet(s3TagSet map[string]string, infraTagSet map[string]string) (tagUpdatedCount int) {
+	for key, value := range infraTagSet {
+		// If a tag is value is empty, it's marked for deletion
+		// and is deleted from the list obtained from S3 bucket
+		if value == "" {
+			klog.V(5).Infof("%s tag will be deleted", key)
+			delete(s3TagSet, key)
+			tagUpdatedCount++
+			continue
+		}
+		val, ok := s3TagSet[key]
+		if !ok || val != value {
+			klog.V(5).Infof("%s tag will be added/updated with value %s", key, value)
+			s3TagSet[key] = value
+			tagUpdatedCount++
+		}
+	}
+	return
+}
```

Comment:
I'm a bit surprised to see that this function actually updates `s3TagSet`. The name does not indicate any potential mutation to its parameters. It would be more clear if this function would 1) not mutate the parameters; and 2) be more explicitly named.

For the parameter mutation, we could simply use a new map instead of mutating `s3TagSet`. 
Here's what it could look like:
```suggestion
// syncInfraTags synchronizes the tags obtained from S3 bucket and Infrastructure CR.
// Returns a new map containing the new tag set, omitting tags that were deleted
// from the infrastructure CR, and adding or updating tags that are new or have been modified.
func syncInfraTags(s3TagSet map[string]string, infraTagSet map[string]string) (newTagSet map[string]string, tagUpdatedCount int) {
	for key, value := range infraTagSet {
		// If a tag is value is empty, it's marked for deletion,
		// so we omit it from the new tag set.
		if value == "" {
			klog.V(5).Infof("%s tag will be deleted", key)
			tagUpdatedCount++
			continue
		}
		val, ok := s3TagSet[key]
		if !ok || val != value {
			klog.V(5).Infof("%s tag will be added/updated with value %s", key, value)
			newTagSet[key] = value
			tagUpdatedCount++
		}
	}
	return
}
```

The caller of this function would also need to change if you adopt this suggestion so that uses the returned map to update the S3 tags.

WDYT?

Reply:
I updated the function name and comments. 
So basically I am following the way upstream CAPI has the way to Merge the [tags](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/c2c96dd3d11d57cd2dfa4faaf28f7476c4a2a005/api/v1beta2/tags.go#L65)